core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```diff
@@ -536,11 +536,15 @@ private[spark] class IndexShuffleBlockResolver(
       dirs: Option[Array[String]] = None): File = {
     val blockId = ShuffleChecksumBlockId(shuffleId, mapId, NOOP_REDUCE_ID)
     val fileName = ShuffleChecksumHelper.getChecksumFileName(blockId, conf)
-    dirs
-      .map(ExecutorDiskUtils.getFile(_, blockManager.subDirsPerLocalDir, fileName))
-      .getOrElse {
-        blockManager.diskBlockManager.getFile(fileName)
-      }
+    // We should use the blockId.name as the file name first to create the file so that
+    // readers (e.g., shuffle external service) without knowing the checksum algorithm
+    // could also find the file.
+    val file = dirs
+      .map(ExecutorDiskUtils.getFile(_, blockManager.subDirsPerLocalDir, blockId.name))
+      .getOrElse(blockManager.diskBlockManager.getFile(blockId))
```

Comment:
Here, we are assuming how `getChecksumFileName` constructs the `fileName` in order to get to the block's filename here - any change there will need to be duplicated here.
Instead, why not simply strip the suffix and use that ?


```suggestion
    val fileNameWithoutChecksum = fileName.substring(0, fileName.lastIndexOf('.'))
    val file = dirs
      .map(ExecutorDiskUtils.getFile(_, blockManager.subDirsPerLocalDir, fileNameWithoutChecksum))
      .getOrElse(blockManager.diskBlockManager.getFile(fileNameWithoutChecksum))
```